pkg/controllers/management/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -497,116 +443,82 @@ func (m *MCAppManager) deleteApps(mcAppName string, mcapp *v3.MultiClusterApp) (
 		return nil, err
 	}
 	return nil, nil
 }
 
-func (m *MCAppManager) getAllApps(mcAppName string, withContext bool) ([]*pv3.App, map[string]string, map[string]*config.UserContext, error) {
+func (m *MCAppManager) getAllApps(mcAppName string) ([]*pv3.App, error) {
 	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
 	allApps := []*pv3.App{}
-	projectToCluster := map[string]string{}
-	clusterContexts := map[string]*config.UserContext{}
 	set := labels.Set(map[string]string{MultiClusterAppIDSelector: mcAppName})
 	clusters, err := m.clusterLister.List("", labels.NewSelector())
 	if err != nil {
-		return allApps, projectToCluster, clusterContexts, err
+		return allApps, err
 	}
 	for _, c := range clusters {
 		projects, err := m.projectLister.List(c.Name, labels.NewSelector())
 		if err != nil {
-			return allApps, projectToCluster, clusterContexts, err
+			return allApps, err
 		}
 		for _, p := range projects {
 			apps, err := m.appLister.List(p.Name, set.AsSelector())
 			if err != nil {
-				return allApps, projectToCluster, clusterContexts, err
+				return allApps, err
 			}
 			allApps = append(allApps, apps...)
-			projectToCluster[p.Name] = c.Name
-		}
-		if withContext {
-			userContext, err := m.clusterManager.UserContext(c.Name)
-			if err != nil {
-				return allApps, projectToCluster, clusterContexts, err
-			}
-			clusterContexts[c.Name] = userContext
 		}
 	}
-	return allApps, projectToCluster, clusterContexts, err
+	return allApps, err
 }
 
-func (m *MCAppManager) reconcileTargetsForDelete(mcapp *v3.MultiClusterApp, clusterContexts map[string]*config.UserContext) error {
-	existingNS := map[string]bool{}
-	for ind, t := range mcapp.Spec.Targets {
+func (m *MCAppManager) reconcileTargetsForDelete(mcapp *v3.MultiClusterApp) error {
+	existingApps := map[string]bool{}
+	set := labels.Set(map[string]string{MultiClusterAppIDSelector: mcapp.Name})
+	for _, t := range mcapp.Spec.Targets {
 		split := strings.SplitN(t.ProjectName, ":", 2)
 		if len(split) != 2 {
 			return fmt.Errorf("error in splitting project ID %v", t.ProjectName)
 		}
 		projectNS := split[1]
-		nsName := getAppNamespaceName(mcapp.Name, projectNS)
-		ns, err := clusterContexts[split[0]].Core.Namespaces("").Controller().Lister().Get("", nsName)
+		apps, err := m.appLister.List(projectNS, set.AsSelector())
 		if err != nil {
-			if !apierrors.IsNotFound(err) {
-				return err
-			}
+			return err
 		}
-		if ns == nil {
-			mcapp.Spec.Targets[ind].AppName = ""
-		} else {
-			existingNS[ns.Name] = true
+		for _, app := range apps {
+			existingApps[app.Namespace] = true
 		}
 	}
-	allApps, projectToCluster, _, err := m.getAllApps(mcapp.Name, false)
+	allApps, err := m.getAllApps(mcapp.Name)
 	if err != nil {
 		return err
 	}
 	toDelete := []*pv3.App{}
 	for _, app := range allApps {
-		if _, ok := existingNS[app.Spec.TargetNamespace]; !ok {
+		if _, ok := existingApps[app.Namespace]; !ok {
 			toDelete = append(toDelete, app)
```

Comment:
Is it possible that getAllApps could return more than 1 app in the same project with the multiclusterapp label, and since both apps will have the same app.Namespace, we don't delete anything?

Reply:
Discussed offline and more than one apps being in a project for the same multilcusterapp won't be an issue since app name is mcapp.Name + project name

Reply:
Discussed offline, there'll be one app per mcApp in a project, so we think it should be okay